pravega/pravega
# Issue 3319: Micrometer metrics migration

Signed-off-by: kevinhan88 <kevinhan88@gmail.com>

**Change log description**  
Changed to micrometer based implementation to support dimensional data model.

**Purpose of the change**  
Fixes #3319 

**What the code does**  
Replaced the entire underlying metric libraries to micrometer, and refactored all the metric classes and test codes.
. Added Micrometer core, statsd and influx libraries.
. Added varArgument "String... tags" to all the metric lookup APIs
. Added InfluxDB config so that Pravega can talk to InfluxDB directly.
. Added "nameFromTags" utility method to form metric lookup key using tags
. Removed dec() from Counter API since it is obviously anti-principal anti-pattern.
. Replaced codahale system metrics with Micrometer's implementation
. Replaced getName() of Metric base class with getId(), because Codahale lookup is String based but Micrometer lookup is Id based.
. Align OpStatsDataPERCENTILESET with Micrometer's percentile set.
. Refactored CounterImpl, GaugeImpl, MeterImpl and TimerImpl to be Micrometer based.
. Fixed FileSystemStorageTest as it didn't bind the metric register to a concrete storage. 

**What the code does not do**
All the essential API changes have been baked in this PR, but no tags have been added into metrics.
Following this PR, there is going to be another PR to make all the metrics changes, initially based on the telegraf.conf provided by the Platform team.
For example, for metric "container.failover" with dimension "container", today we call nameFromContainer("container.failover", 6) to construct a flat metric name "container.failover.6", and then call incCounterValue("container.failover.6", 1) - with this PR everything works as before. 

With the next PR, we no longer construct the flat name, instead we call incCounterValue("container.failover", 1, "containerId", "6") - note the tag pair(s) added.

**How to verify it**  
This is the first step of the refactoring, which is to replace the underlying libraries.
Because the change is backwards compatible, all the existing hierarchical metrics should still work.
After step one, telegraf agent should work the same way as before. However it is not ready to talk to InfluxDB directly.

The second step is to add tags into the required metrics (according to the Platform team's mapping file). 
After this step, the metric should be able to talk directly to InfluxDB and Telegraf agent.
